File: java/src/test/java/com/squirrelparser/ErrorLocalizationTest.java
// ===========================================================================
// ERROR LOCALIZATION TESTS (Non-Cascading Verification)
// ===========================================================================
// These tests verify that errors don't cascade - each error is localized
// to its specific location without affecting subsequent parsing.

package com.squirrelparser;

import org.junit.jupiter.api.Test;

import static com.squirrelparser.TestUtils.*;
import static org.junit.jupiter.api.Assertions.*;

class ErrorLocalizationTest {

    @Test
    void cascade01_errorInFirstElementDoesntAffectSecond() {
        // Error in first element, second element parses cleanly
        String grammar = "S <- \"a\" \"b\" \"c\" ;";
        var result = testParse(grammar, "aXbc");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have exactly 1 error (at position 1)");
        assertTrue(result.skippedStrings().contains("X"), "should skip X");
        // Error localized to position 1, doesn't cascade to 'b' or 'c'
    }

    @Test
    void cascade02_errorInNestedStructure() {
        // Error inside inner Seq, doesn't affect outer Seq
        String grammar = """
            S <- ("a" "b") "c" ;
        """;
        var result = testParse(grammar, "aXbc");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have exactly 1 error");
        assertTrue(result.skippedStrings().contains("X"), "should skip X");
        // Error in inner Seq (between 'a' and 'b'), outer Seq continues normally
    }

    @Test
    void cascade03_lrErrorDoesntCascadeToNextExpansion() {
        // Error in one LR expansion iteration, next iteration clean
        String grammar = """
            E <- E "+" "n" / "n" ;
        """;
        var result = testParse(grammar, "n+Xn+n", "E");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have exactly 1 error");
        assertTrue(result.skippedStrings().contains("X"), "should skip X");
        // Expansion: n (base), n+[skip X]n, n+Xn+n
        // First '+' clean, second '+' has error, third '+' clean
        // Error localized to second iteration
    }

    @Test
    void cascade04_multipleIndependentErrors() {
        // Multiple errors in different parts of parse, all localized
        String grammar = """
            S <- ("a" "b") ("c" "d") ("e" "f") ;
        """;
        var result = testParse(grammar, "aXbcYdeZf");
        assertTrue(result.ok(), "should succeed");
        assertEquals(3, result.errorCount(), "should have 3 independent errors");
        assertTrue(result.skippedStrings().contains("X"), "should skip X");
        assertTrue(result.skippedStrings().contains("Y"), "should skip Y");
        assertTrue(result.skippedStrings().contains("Z"), "should skip Z");
        // Each error localized to its respective Seq
    }

    @Test
    void cascade05_errorBeforeRepetition() {
        // Error before repetition, repetition parses cleanly
        String grammar = """
            S <- "a" "b"+ ;
        """;
        var result = testParse(grammar, "aXbbb");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error");
        assertTrue(result.skippedStrings().contains("X"), "should skip X");
        // Error at position 1, OneOrMore starts cleanly at position 2
    }

    @Test
    void cascade06_errorAfterRepetition() {
        // Repetition clean, error after it
        String grammar = """
            S <- "a"+ "b" ;
        """;
        var result = testParse(grammar, "aaaXb");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error");
        assertTrue(result.skippedStrings().contains("X"), "should skip X");
        // OneOrMore matches 3 a's cleanly, then error, then 'b'
    }

    @Test
    void cascade07_errorInFirstAlternativeDoesntPoisonSecond() {
        // First alternative has error, second alternative clean
        String grammar = """
            S <- "a" "b" / "c" ;
        """;
        var result = testParse(grammar, "c");
        assertTrue(result.ok(), "should succeed");
        assertEquals(0, result.errorCount(), "should have 0 errors (second alternative clean)");
        // First tries and fails, second succeeds cleanly - no cascade
    }

    @Test
    void cascade08_recoveryVersionIncrementsCorrectly() {
        // Each recovery increments version, ensuring proper cache invalidation
        String grammar = """
            S <- ("a" "b") ("c" "d") ;
        """;
        var result = testParse(grammar, "aXbcYd");
        assertTrue(result.ok(), "should succeed");
        assertEquals(2, result.errorCount(), "should have 2 errors");
        // Two recoveries, each increments version, no cache pollution
    }

    @Test
    void cascade09_errorAtDeeplyNestedLevel() {
        // Error very deep in nesting, doesn't affect outer levels
        String grammar = """
            S <- ((("a" "b") "c") "d") "e" ;
        """;
        var result = testParse(grammar, "aXbcde");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error at deepest level");
        assertTrue(result.skippedStrings().contains("X"), "should skip X");
        // Error localized despite 4 levels of nesting
    }

    @Test
    void cascade10_errorRecoveryDoesntLeaveParserInBadState() {
        // After recovery, parser continues with clean state
        String grammar = """
            S <- ("a" "b") "c" ("d" "e") ;
        """;
        var result = testParse(grammar, "abXcde");
        assertTrue(result.ok(), "should succeed");
        assertEquals(1, result.errorCount(), "should have 1 error");
        // After skipping X at position 2, matches 'c' at position 3, then 'de'
    }
}

File: dart/test/parser/error_localization_test.dart
// ===========================================================================
// ERROR LOCALIZATION TESTS (Non-Cascading Verification)
// ===========================================================================
// These tests verify that errors don't cascade - each error is localized
// to its specific location without affecting subsequent parsing.

import 'package:test/test.dart';

import '../test_utils.dart';

void main() {
  group('Error Localization (Non-Cascading) Tests', () {
    test('CASCADE-01-error-in-first-element-doesnt-affect-second', () {
      // Error in first element, second element parses cleanly
      const grammar = 'S <- "a" "b" "c" ;';
      final (ok, err, skip) = testParse(grammar, 'aXbc');
      expect(ok, isTrue, reason: 'should succeed');
      expect(err, equals(1),
          reason: 'should have exactly 1 error (at position 1)');
      expect(skip.contains('X'), isTrue, reason: 'should skip X');
      // Error localized to position 1, doesn't cascade to 'b' or 'c'
    });

    test('CASCADE-02-error-in-nested-structure', () {
      // Error inside inner Seq, doesn't affect outer Seq
      const grammar = '''
        S <- ("a" "b") "c" ;
      ''';
      final (ok, err, skip) = testParse(grammar, 'aXbc');
      expect(ok, isTrue, reason: 'should succeed');
      expect(err, equals(1), reason: 'should have exactly 1 error');
      expect(skip.contains('X'), isTrue, reason: 'should skip X');
      // Error in inner Seq (between 'a' and 'b'), outer Seq continues normally
    });

    test('CASCADE-03-lr-error-doesnt-cascade-to-next-expansion', () {
      // Error in one LR expansion iteration, next iteration clean
      const grammar = '''
        E <- E "+" "n" / "n" ;
      ''';
      final (ok, err, skip) = testParse(grammar, 'n+Xn+n', 'E');
      expect(ok, isTrue, reason: 'should succeed');
      expect(err, equals(1), reason: 'should have exactly 1 error');
      expect(skip.contains('X'), isTrue, reason: 'should skip X');
      // Expansion: n (base), n+[skip X]n, n+Xn+n
      // First '+' clean, second '+' has error, third '+' clean
      // Error localized to second iteration
    });

    test('CASCADE-04-multiple-independent-errors', () {
      // Multiple errors in different parts of parse, all localized
      const grammar = '''
        S <- ("a" "b") ("c" "d") ("e" "f") ;
      ''';
      final (ok, err, skip) = testParse(grammar, 'aXbcYdeZf');
      expect(ok, isTrue, reason: 'should succeed');
      expect(err, equals(3), reason: 'should have 3 independent errors');
      expect(skip.contains('X'), isTrue, reason: 'should skip X');
      expect(skip.contains('Y'), isTrue, reason: 'should skip Y');
      expect(skip.contains('Z'), isTrue, reason: 'should skip Z');
      // Each error localized to its respective Seq
    });

    test('CASCADE-05-error-before-repetition', () {
      // Error before repetition, repetition parses cleanly
      const grammar = '''
        S <- "a" "b"+ ;
      ''';
      final (ok, err, skip) = testParse(grammar, 'aXbbb');
      expect(ok, isTrue, reason: 'should succeed');
      expect(err, equals(1), reason: 'should have 1 error');
      expect(skip.contains('X'), isTrue, reason: 'should skip X');
      // Error at position 1, OneOrMore starts cleanly at position 2
    });

    test('CASCADE-06-error-after-repetition', () {
      // Repetition clean, error after it
      const grammar = '''
        S <- "a"+ "b" ;
      ''';
      final (ok, err, skip) = testParse(grammar, 'aaaXb');
      expect(ok, isTrue, reason: 'should succeed');
      expect(err, equals(1), reason: 'should have 1 error');
      expect(skip.contains('X'), isTrue, reason: 'should skip X');
      // OneOrMore matches 3 a's cleanly, then error, then 'b'
    });

    test('CASCADE-07-error-in-first-alternative-doesnt-poison-second', () {
      // First alternative has error, second alternative clean
      const grammar = '''
        S <- "a" "b" / "c" ;
      ''';
      final (ok, err, _) = testParse(grammar, 'c');
      expect(ok, isTrue, reason: 'should succeed');
      expect(err, equals(0),
          reason: 'should have 0 errors (second alternative clean)');
      // First tries and fails, second succeeds cleanly - no cascade
    });

    test('CASCADE-08-recovery-version-increments-correctly', () {
      // Each recovery increments version, ensuring proper cache invalidation
      const grammar = '''
        S <- ("a" "b") ("c" "d") ;
      ''';
      final (ok, err, _) = testParse(grammar, 'aXbcYd');
      expect(ok, isTrue, reason: 'should succeed');
      expect(err, equals(2), reason: 'should have 2 errors');
      // Two recoveries, each increments version, no cache pollution
    });

    test('CASCADE-09-error-at-deeply-nested-level', () {
      // Error very deep in nesting, doesn't affect outer levels
      const grammar = '''
        S <- ((("a" "b") "c") "d") "e" ;
      ''';
      final (ok, err, skip) = testParse(grammar, 'aXbcde');
      expect(ok, isTrue, reason: 'should succeed');
      expect(err, equals(1), reason: 'should have 1 error at deepest level');
      expect(skip.contains('X'), isTrue, reason: 'should skip X');
      // Error localized despite 4 levels of nesting
    });

    test('CASCADE-10-error-recovery-doesnt-leave-parser-in-bad-state', () {
      // After recovery, parser continues with clean state
      const grammar = '''
        S <- ("a" "b") "c" ("d" "e") ;
      ''';
      final (ok, err, _) = testParse(grammar, 'abXcde');
      expect(ok, isTrue, reason: 'should succeed');
      expect(err, equals(1), reason: 'should have 1 error');
      // After skipping X at position 2, matches 'c' at position 3, then 'de'
    });
  });
}
